[inductor] Pass None and skip constexpr in custom Triton kernel calls from C++

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114475

Summary: `None` arguments are codegened as `*i8` in the `triton_meta` of the generated or user-defined Triton kernels:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/triton_utils.py#L33-L36

Due to this, in contrary to the conventional Triton, we actually should pass `nullptr` to the Triton kernels in C++ wrapper codegen instead of passing nothing (as normally `None` doesn't make it to the generated PTX parameters, just like `tl.constexpr` args).

This PR adds two things:

1. Proper C++ wrapper codegening (ABI and non-ABI) of `nullptr` and `c10::nullopt`, as the prior way codegening `c10::nullopt` as tensor breaks (also `c10` breaks in the ABI mode).

2. Skipping `tl.constexpr` args when calling the loaded-from-cubin compiled Triton kernel in the C++ wrapper codegen. As a side effect, this also resolves an issue with string arguments: now they are simply omitted in the C++ wrapper codegen.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_triton_kernel_with_none_input
...
----------------------------------------------------------------------
Ran 4 tests in 40.364s

OK (skipped=2)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler

aakhundov committed:
Pass None and skip constexpr in custom Triton kernel calls from C++

Summary: `None` arguments are codegened as `*i8` in the `triton_meta` of the generated or user-defined Triton kernels:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/triton_utils.py#L33-L36

Due to this, in contrary to the conventional Triton, we actually should pass `nullptr` to the Triton kernels in C++ wrapper codegen instead of passing nothing (as normally `None` doesn't make it to the generated PTX parameters, just like `tl.constexpr` args).

This PR adds two things:

1. Proper C++ wrapper codegening (ABI and non-ABI) of `nullptr` and `c10::nullopt`, as the prior way codegening `c10::nullopt` as tensor breaks (also `c10` breaks in the ABI mode).

2. Skipping `tl.constexpr` args when calling the loaded-from-cubin compiled Triton kernel in the C++ wrapper codegen. As a side effect, this also resolves an issue with string arguments: now they are simply omitted in the C++ wrapper codegen.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_triton_kernel_with_none_input
...
----------------------------------------------------------------------
Ran 4 tests in 40.364s

OK (skipped=2)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

oulgen commented on test/inductor/test_aot_inductor.py:
```diff
@@ -1268,6 +1269,50 @@ def forward(self, x):
+
+                wo_kernel = add_kernel_with_optional_param[(1,)](
+                    x,
+                    None,
```
I thought the problem was it was being passed a tensor first and then later None?

aakhundov replied:
This ended up being more subtle than we thought. Tensor arguments can be `None` or not `None`, bu (I believe) checking if `in_ptr1 is None` in the Triton kernel code doesn't quite work: the generated PTX code doesn't seem to be affected by this. Instead, there should be another, typically a `tl.constexpr` argument that decides the control flow, set in accordance with the respective tensor arguments being set to `None` or not. As a result, your existing de-duping mechanism doesn't dedup the calls with `None` and not, due to another constant arg getting different values.

oulgen replied:
Ok, so de-duping mechanism is correct then

aakhundov replied:
It's not necessarily guarded against None / non-None tensor args, but those are normally accompanied by constexpr args with different values, so should be fine.

oulgen commented on torch/_inductor/ir.py:
```diff
@@ -3830,13 +3830,23 @@ def codegen(self, wrapper):
+            # get added asparameters to the PTX code compiled from the
+            # user-defined Triton kernel (only non-constexpr args do)
+            kernel = kernel_side_table.get_kernel(self.kernel_idx)
+            args = [arg for i, arg in enumerate(args) if i not in kernel.constexprs]
```
I am not sure whether this is guaranteed to be correct. Your 'args' comes from kwargs, which are not guaranteed to be the right order, so I think you need to sort them to the correct order expected by the kernel. I did this somewhere else for the C++ codegen as well.

aakhundov replied:
Before this PR, in the CPP wrapper mode, `self.codegen_kwargs()` used to be passed directly to the `CppWrapperCodeGen.generate_user_defined_triton_kernel` method here:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/wrapper.py#L1771

which, in turn, passes the `args`, without modifying the order, to the `CppWrapperCodeGen.generate_kernel_call` method here:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/wrapper.py#L2570-L2572

finally, the `WrapperCodeGen.wrapper_kernel_call` is called:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/wrapper.py#L980

Looking into the chain of calls above, I don't see reordering of the args along the code path. Am I missing something?

Importantly, when compiling a Triton kernel and skipping the `tl.constexpr` parameters, Triton compiler seems to preserve the order of the remaining ones in the emitted PTX code.

aakhundov replied:
I think, similar thing is done here for similar purposes?

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/triton_heuristics.py#L310-L314

oulgen replied:
`self.fn.arg_names` are already sorted becayse they come from the definition

aakhundov replied:
@oulgen Ok, I think the ordering you're talking about is already happening in `self.codegen_kwargs`:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/ir.py#L3636-L3642

So we should be good.

oulgen replied:
For the C++ wrapper, the sorting happen on
https://github.com/pytorch/pytorch/blob/fd1a01a393cfe26205795a7fae48ff4e49b5a564/torch/_inductor/ir.py#L3637-L3642

which is right before this, so i think we are all good since you're operating over already sorted list

aakhundov committed:
Update on "[inductor] Pass None and skip constexpr in custom Triton kernel calls from C++"


Summary: `None` arguments are codegened as `*i8` in the `triton_meta` of the generated or user-defined Triton kernels:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/triton_utils.py#L33-L36

Due to this, in contrary to the conventional Triton, we actually should pass `nullptr` to the Triton kernels in C++ wrapper codegen instead of passing nothing (as normally `None` doesn't make it to the generated PTX parameters, just like `tl.constexpr` args).

This PR adds two things:

1. Proper C++ wrapper codegening (ABI and non-ABI) of `nullptr` and `c10::nullopt`, as the prior way codegening `c10::nullopt` as tensor breaks (also `c10` breaks in the ABI mode).

2. Skipping `tl.constexpr` args when calling the loaded-from-cubin compiled Triton kernel in the C++ wrapper codegen. As a side effect, this also resolves an issue with string arguments: now they are simply omitted in the C++ wrapper codegen.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_triton_kernel_with_none_input
...
----------------------------------------------------------------------
Ran 4 tests in 40.364s

OK (skipped=2)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov committed:
Update on "[inductor] Pass None and skip constexpr in custom Triton kernel calls from C++"


Summary: `None` arguments are codegened as `*i8` in the `triton_meta` of the generated or user-defined Triton kernels:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/triton_utils.py#L33-L36

Due to this, in contrary to the conventional Triton, we actually should pass `nullptr` to the Triton kernels in C++ wrapper codegen instead of passing nothing (as normally `None` doesn't make it to the generated PTX parameters, just like `tl.constexpr` args).

This PR adds two things:

1. Proper C++ wrapper codegening (ABI and non-ABI) of `nullptr` and `c10::nullopt`, as the prior way codegening `c10::nullopt` as tensor breaks (also `c10` breaks in the ABI mode).

2. Skipping `tl.constexpr` args when calling the loaded-from-cubin compiled Triton kernel in the C++ wrapper codegen. As a side effect, this also resolves an issue with string arguments: now they are simply omitted in the C++ wrapper codegen.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_triton_kernel_with_none_input
...
----------------------------------------------------------------------
Ran 4 tests in 40.364s

OK (skipped=2)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

oulgen commented on torch/_inductor/ir.py:
```diff
@@ -3830,13 +3843,20 @@ def codegen(self, wrapper):
+            # in C++ wrapper, we don't pass constexpr args, as they don't
+            # get added as parameters to the PTX code compiled from the
+            # user-defined Triton kernel (only non-constexpr args do)
+            args = self._remove_constexpr_args(args)
```
The above
```
kernel, configs = self.get_kernel_and_configs()
```
already extracted the kernel for you, so just kernel to this function instead of duplicating that logic

oulgen commented on torch/_inductor/ir.py:
```diff
@@ -3810,6 +3810,19 @@ def apply_constraint(self):
 
 
 class UserDefinedTritonKernel(ExternKernel):
+    def _remove_constexpr_args(self, args):
```
take kernel here from below and make this a static function

you can in fact inline this to codegen since you dont call this function anywhere else

oulgen replied:
Actually after taking the extracted kernel in def codegen, this function is only one line now, please just inline it below instead, no need for new function

aakhundov replied:
Few lines, but inlined.

aakhundov committed:
Update on "[inductor] Pass None and skip constexpr in custom Triton kernel calls from C++"


Summary: `None` arguments are codegened as `*i8` in the `triton_meta` of the generated or user-defined Triton kernels:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/triton_utils.py#L33-L36

Due to this, in contrary to the conventional Triton, we actually should pass `nullptr` to the Triton kernels in C++ wrapper codegen instead of passing nothing (as normally `None` doesn't make it to the generated PTX parameters, just like `tl.constexpr` args).

This PR adds two things:

1. Proper C++ wrapper codegening (ABI and non-ABI) of `nullptr` and `c10::nullopt`, as the prior way codegening `c10::nullopt` as tensor breaks (also `c10` breaks in the ABI mode).

2. Skipping `tl.constexpr` args when calling the loaded-from-cubin compiled Triton kernel in the C++ wrapper codegen. As a side effect, this also resolves an issue with string arguments: now they are simply omitted in the C++ wrapper codegen.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_triton_kernel_with_none_input
...
----------------------------------------------------------------------
Ran 4 tests in 40.364s

OK (skipped=2)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

oulgen commented on torch/_inductor/ir.py:
```diff
@@ -3848,7 +3835,13 @@ def codegen(self, wrapper):
             # get added as parameters to the PTX code compiled from the
             # user-defined Triton kernel (only non-constexpr args do)
-            args = self._remove_constexpr_args(args)
+            from triton.runtime.jit import JITFunction
```
You dont need any of this, it is already a JITFunction

kernel, configs = self.get_kernel_and_configs()

took care of it for you

oulgen replied:
3839-3843 are not needed

aakhundov replied:
Nope: autotuning tests fail. There `kernel` is an `Autotuner` instance which doesn't have `constexpr`. Need to unwrap. That's why I updated the PR.

aakhundov committed:
Update on "[inductor] Pass None and skip constexpr in custom Triton kernel calls from C++"


Summary: `None` arguments are codegened as `*i8` in the `triton_meta` of the generated or user-defined Triton kernels:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/triton_utils.py#L33-L36

Due to this, in contrary to the conventional Triton, we actually should pass `nullptr` to the Triton kernels in C++ wrapper codegen instead of passing nothing (as normally `None` doesn't make it to the generated PTX parameters, just like `tl.constexpr` args).

This PR adds two things:

1. Proper C++ wrapper codegening (ABI and non-ABI) of `nullptr` and `c10::nullopt`, as the prior way codegening `c10::nullopt` as tensor breaks (also `c10` breaks in the ABI mode).

2. Skipping `tl.constexpr` args when calling the loaded-from-cubin compiled Triton kernel in the C++ wrapper codegen. As a side effect, this also resolves an issue with string arguments: now they are simply omitted in the C++ wrapper codegen.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_triton_kernel_with_none_input
...
----------------------------------------------------------------------
Ran 4 tests in 40.364s

OK (skipped=2)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

oulgen commented on torch/_inductor/ir.py:
```diff
@@ -3830,13 +3830,26 @@ def codegen(self, wrapper):
+            from triton.runtime.autotuner import Autotuner
+            from triton.runtime.jit import JITFunction
+
+            if isinstance(kernel, Autotuner):
+                kernel = kernel.fn
```
I'm still a bit confused about your comment there.
```
kernel, configs = self.get_kernel_and_configs()
```
is guaranteed to return a kernel such that
```
assert isinstance(kernel, JITFunction)
```
if you look at line 3820 above, it does kernel = kernel.fn when it is autotuner.

aakhundov replied:
Yeah, I think, you're right. The autotuning test has failed before when I was fetching the `kernel` myself. Now that I'm reusing the `kernel` returned by `get_kernel_and_configs`, I can skip the unwrapping as it's already done.

aakhundov committed:
Update on "[inductor] Pass None and skip constexpr in custom Triton kernel calls from C++"


Summary: `None` arguments are codegened as `*i8` in the `triton_meta` of the generated or user-defined Triton kernels:

https://github.com/pytorch/pytorch/blob/85aa3723749e0d06aa5fd34215b9b93529a60995/torch/_inductor/codegen/triton_utils.py#L33-L36

Due to this, in contrary to the conventional Triton, we actually should pass `nullptr` to the Triton kernels in C++ wrapper codegen instead of passing nothing (as normally `None` doesn't make it to the generated PTX parameters, just like `tl.constexpr` args).

This PR adds two things:

1. Proper C++ wrapper codegening (ABI and non-ABI) of `nullptr` and `c10::nullopt`, as the prior way codegening `c10::nullopt` as tensor breaks (also `c10` breaks in the ABI mode).

2. Skipping `tl.constexpr` args when calling the loaded-from-cubin compiled Triton kernel in the C++ wrapper codegen. As a side effect, this also resolves an issue with string arguments: now they are simply omitted in the C++ wrapper codegen.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_triton_kernel_with_none_input
...
----------------------------------------------------------------------
Ran 4 tests in 40.364s

OK (skipped=2)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge
